src/radical/entk/appman/wfprocessor.py

```diff
@@ -140,105 +147,119 @@ def _enqueue(self, local_prof):
 
                     with pipe.lock:
 
-                        if ((not pipe.completed) and (not pipe.state == states.SUSPENDED)):
+                        # Test if the pipeline is already in the final state
+                        if  pipe.state in states.FINAL    or \
+                            pipe.completed                or \
+                            pipe.state == states.SUSPENDED:
+                            continue
+
+                        if pipe.state == states.INITIAL:
+
+                            # Set state of pipeline to SCHEDULING if it is in INITIAL
+                            transition(obj=pipe,
+                                       obj_type='Pipeline',
+                                       new_state=states.SCHEDULING,
+                                       channel=mq_channel,
+                                       queue='%s-enq-to-sync' % self._sid,
+                                       profiler=local_prof,
+                                       logger=self._logger)
+
+                        executable_stage = pipe.stages[pipe.current_stage - 1]
+
+                        if not executable_stage.uid:
+                            # AM: why are pipe UID's assigned here, but not via
+                            # assign_uid?
```

Comment:
The point I am trying to make is that a method exists for that (`assign_uid`) - why is it not used?

Reply:
Because, as far as I understand the code, `assign_uid` method provides `uids` to pipelines, stages and tasks when have not have a uid yet.

The pipeline already has its `uid` and the stage has its `uid`. It takes the pipeline's `uid` and assigns it to the stage `parent_uid`. I do not see why a new `uid` should be created.

@vivek-bala, what is actually happening?